### PR TITLE
Test flakes: use ordering tests for keep alives

### LINKED ISSUE #5346: Missing keep-alive in lib/backend
Keep-alives disappear out-of-band.

<details><summary><code>Logs</code></summary>

----------------------------------------------------------------------
FAIL: etcd_test.go:127: EtcdSuite.TestKeepAlive

etcd_test.go:128:
    s.suite.KeepAlive(c)
/go/src/github.com/gravitational/teleport/lib/backend/test/suite.go:345:
    c.Assert(err, check.IsNil)
... value *trace.TraceErr = 
ERROR REPORT:
Original Error: *trace.NotFoundError item &#34;/63f9ce2d-7c13-402f-a2f6-b0069f58242bkey&#34; is not found
Stack Trace:
	/go/src/github.com/gravitational/teleport/lib/backend/etcdbk/etcd.go:586 github.com/gravitational/teleport/lib/backend/etcdbk.(*EtcdBackend).Get
	/go/src/github.com/gravitational/teleport/lib/backend/test/suite.go:344 github.com/gravitational/teleport/lib/backend/test.(*BackendSuite).KeepAlive
	/go/src/github.com/gravitational/teleport/lib/backend/etcdbk/etcd_test.go:128 github.com/gravitational/teleport/lib/backend/etcdbk.(*EtcdSuite).TestKeepAlive
	/opt/go/src/reflect/value.go:476 reflect.Value.call
	/opt/go/src/reflect/value.go:337 reflect.Value.Call
	/go/src/github.com/gravitational/teleport/vendor/gopkg.in/check.v1/check.go:781 gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1
	/go/src/github.com/gravitational/teleport/vendor/gopkg.in/check.v1/check.go:675 gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1
	/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit
User Message: item &#34;/63f9ce2d-7c13-402f-a2f6-b0069f58242bkey&#34; is not found ("item \"/63f9ce2d-7c13-402f-a2f6-b0069f58242bkey\" is not found")

OOPS: 10 passed, 1 FAILED
--- FAIL: TestEtcd (9.73s)
{"level":"warn","ts":"2021-01-12T00:35:14.440Z","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-e413cc2c-d3f2-40f2-a45c-27d2703f1391/127.0.0.1:2379","attempt":0,"error":"rpc error: code = ResourceExhausted desc = trying to send message larger than max (293 vs. 128)"}
FAIL
coverage: 77.3% of statements
FAIL	github.com/gravitational/teleport/lib/backend/etcdbk	9.913s
```
</details>

### PR DESCRIPTION
**Pull request by a-palchikov**
Evaluate watcher events to decide whether keep-alives are effective instead of relying on arbitrary TTLs.
Exact TTLs are susceptible to skewing when used on a real backend (etcd).

I opted to keep this change private to the `etcd` backend although it may replace the suite's implementation as I've also observed the same failure with the `lite` backend.

Fixes https://github.com/gravitational/teleport/issues/5346.

### CODE REVIEW

**Review by russjones — COMMENTED** [review]
I would make this change to the suite. It will prevent regressions and make implementing new backends easier.

File: lib/auth/apiserver_test.go [hunk 1]
```diff
@@ -129,7 +131,7 @@ func TestUpsertServer(t *testing.T) {
 			addServers(s.GetAuthServers())
 			addServers(s.GetNodes(defaults.Namespace))
 			addServers(s.GetProxies())
-			require.Empty(t, cmp.Diff(allServers, []services.Server{tt.wantServer}))
+			require.Empty(t, cmp.Diff(allServers, []services.Server{tt.wantServer}, cmpopts.IgnoreFields(types.Metadata{}, "ID")))
```
Comment (a-palchikov) [inline lib/auth/apiserver_test.go:134]: Due to changes to memory backend, now it behaves similar to etcd and generates new IDs in `PutXXX` (see below).

**Comment by a-palchikov** [comment]
I moved the keep alive changes to test/suite as @russjones suggested.

File: lib/backend/test/suite.go [hunk 2]
```diff
@@ -322,52 +321,39 @@ func (s *BackendSuite) KeepAlive(c *check.C) {
...
-	c.Assert(events[2].Type, check.Equals, backend.OpPut)
-	c.Assert(string(events[2].Item.Key), check.Equals, string(item.Key))
-
-	c.Assert(string(events[1].Item.Value), check.Equals, string(item.Value))
-	c.Assert(string(events[2].Item.Value), check.Equals, string(item.Value))
-
-	c.Assert(events[1].Item.Expires.IsZero(), check.Not(check.Equals), true, check.Commentf("expected non zero expiration time"))
-	c.Assert(events[2].Item.Expires.IsZero(), check.Not(check.Equals), true, check.Commentf("expected non zero expiration time"))
-	c.Assert(events[2].Item.Expires.After(events[1].Item.Expires), check.Equals, true, check.Commentf("expected %v after %v", events[2].Item.Expires, events[1].Item.Expires))
+	fmt.Println("Recv events:", events)
```
Comment (awly) [inline lib/backend/test/suite.go:340]:
```suggestion
	c.Log("Recv events: ", events)
```

File: lib/backend/test/suite.go [hunk 3]
```diff
@@ -749,3 +752,48 @@ func ExpectItems(c *check.C, items, expected []backend.Item) {
...
+	verifyExpireTimestampsIncreasing(c, obtained, expected)
+}
+
+func verifyIDsIncreasing(c *check.C, obtained []backend.Event) {
+	sorted := make([]backend.Event, len(obtained))
+	copy(sorted, obtained)
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].Item.ID < sorted[j].Item.ID
+	})
+	c.Assert(obtained, check.DeepEquals, sorted)
```
Comment (awly) [inline lib/backend/test/suite.go:768]:
```suggestion
	lastID := int64(-1)
	for _, item := range obtained {
		id := item.Item.ID
		c.Assert(id > lastID, check.Equals, true)
		lastID = id
	}
```

File: lib/backend/test/suite.go [hunk 4]
```diff
@@ -749,3 +752,48 @@ func ExpectItems(c *check.C, items, expected []backend.Item) {
...
+	for id := range dedup {
+		expectedIDs = append(expectedIDs, id)
+	}
+	for _, event := range obtained {
+		obtainedIDs = append(obtainedIDs, event.Item.ID)
+	}
+	sort.Slice(expectedIDs, func(i, j int) bool {
+		return expectedIDs[i] < expectedIDs[j]
+	})
+	c.Assert(obtainedIDs, check.DeepEquals, expectedIDs)
```
Comment (awly) [inline lib/backend/test/suite.go:786]:
```suggestion
	for _, event := range obtained {
		_, ok := dedup[event.Item.ID]
		c.Assert(ok, check.Equals, false)
		dedup[event.Item.ID] = struct{}{}
	}
```

**Review by awly — APPROVED** [review]

File: lib/backend/test/suite.go (hunk 2 above)
Comment (a-palchikov) [reply lib/backend/test/suite.go:340]: That's a debugging left-over - will remove.

**Comment by russjones** [comment]
@fspmarshall Can you review?

**Review by fspmarshall — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**